pandas/core/groupby/groupby.py

```diff
@@ -691,7 +691,14 @@ def _set_result_index_ordered(self, result):
             result.set_axis(index, axis=self.axis, inplace=True)
             result = result.sort_index(axis=self.axis)
 
-        result.set_axis(self.obj._get_axis(self.axis), axis=self.axis, inplace=True)
+        if (
+            self.dropna
+            and self.axis == 0
+            and len(result.index) < len(self._selected_obj.index)
```

Comment:
What happens in the case where dropna is True, len(result.index) < len(self._selected_obj.index), and axis == 1?

Reply:
Another good catch. Groupby+slice with axis=1 throws almost always, it seems, missing values or not. Here are some copy-pastable examples:

With no missing values, slicing down to Series GroupBy, dropna=True works:
```
In [48]: df = pd.DataFrame({"A": [0, 0, 7, 1], "B": [1, 2, 4, 3]}, index=list('abcd')) 
    ...: gb = df.groupby('B', dropna=True, axis=1) 
    ...: gb['B'].transform(len)                                                                                                                                                                                   
Out[48]: Series([], Name: B, dtype: int64)
```
but `dropna=False` throws at the current line:
```
In [47]: df = pd.DataFrame({"A": [0, 0, 7, 1], "B": [1, 2, 4, 3]}, index=list('abcd')) 
    ...: gb = df.groupby('B', dropna=False, axis=1) 
    ...: gb['B'].transform(len)                                                                                                                                                                                   
---------------------------------------------------------------------------
(...)
/workspaces/pandas-arw2019/pandas/core/groupby/groupby.py in _set_result_index_ordered(self, result)
--> 701             result.set_axis(self.obj._get_axis(self.axis), axis=self.axis, inplace=True)
(...)

ValueError: Length mismatch: Expected axis has 2 elements, new values have 4 elements
```
Slicing down to `DataFrameGroupBy` always throws, but the errors are different for `dropna=True` vs `dropna=False`:
```

In [54]: df = pd.DataFrame({"A": [0, 0, 7, 1], "B": [1, 2, 4, 3]}, index=list('abcd')) 
    ...: gb = df.groupby('B', dropna=True, axis=1) 
    ...: gb[['B']].transform(len)                                                                                                                                                                                 
---------------------------------------------------------------------------
(...)
/workspaces/pandas-arw2019/pandas/core/groupby/generic.py in _transform_general(self, func, engine, engine_kwargs, *args, **kwargs)
-> 1358         concatenated = concat(applied, axis=self.axis, verify_integrity=False)
   1359         concatenated = concatenated.reindex(concat_index, axis=other_axis, copy=False)
   1360         return self._set_result_index_ordered(concatenated)
(...)
ValueError: No objects to concatenate
```
```
In [56]: df = pd.DataFrame({"A": [0, 0, 7, 1], "B": [1, 2, 4, 3]}, index=list('abcd')) 
    ...: gb = df.groupby('B', dropna=False, axis=1) 
    ...: gb[['B']].transform(len)                                                                                                                                                                                 
---------------------------------------------------------------------------
(...)
/workspaces/pandas-arw2019/pandas/core/groupby/generic.py in _transform_general(self, func, engine, engine_kwargs, *args, **kwargs)
-> 1311         for name, group in gen:
   1312             object.__setattr__(group, "name", name)

/workspaces/pandas-arw2019/pandas/core/groupby/ops.py in get_iterator(self, data, axis)
    127         splitter = self._get_splitter(data, axis=axis)
    128         keys = self._get_group_keys()
--> 129         for key, (i, group) in zip(keys, splitter):
    130             yield key, group

/workspaces/pandas-arw2019/pandas/core/groupby/ops.py in __iter__(self)
    888     def __iter__(self):
--> 889         sdata = self._get_sorted_data()

/workspaces/pandas-arw2019/pandas/core/groupby/ops.py in _get_sorted_data(self)
    901     def _get_sorted_data(self) -> FrameOrSeries:
--> 902         return self.data.take(self.sort_idx, axis=self.axis)

/workspaces/pandas-arw2019/pandas/core/generic.py in take(self, indices, axis, is_copy, **kwargs)
-> 3540         new_data = self._mgr.take(
   3541             indices, axis=self._get_block_manager_axis(axis), verify=True
   3542         )

/workspaces/pandas-arw2019/pandas/core/internals/managers.py in take(self, indexer, axis, verify, convert)
   1392         n = self.shape[axis]
   1393         if convert:
-> 1394             indexer = maybe_convert_indices(indexer, n)

/workspaces/pandas-arw2019/pandas/core/indexers.py in maybe_convert_indices(indices, n)
    248     mask = (indices >= n) | (indices < 0)
    249     if mask.any():
--> 250         raise IndexError("indices are out-of-bounds")
    251     return indices

IndexError: indices are out-of-bounds
```
With missing values (I'll truncate the tracebacks):
```

In [66]: df = pd.DataFrame({"A": [0, 0, 7, np.nan], "B": [1, 2, 4, np.nan]}, index=list('abcd')) 
    ...: gb = df.groupby('B', dropna=True, axis=1) 
    ...: gb['B'].transform(len)                                                                                                                                                                                   
Out[66]: Series([], Name: B, dtype: float64)

In [67]: df = pd.DataFrame({"A": [0, 0, 7, np.nan], "B": [1, 2, 4, np.nan]}, index=list('abcd')) 
    ...: gb = df.groupby('B', dropna=False, axis=1) 
    ...: gb['B'].transform(len)                                                                                                                                                                                   
---------------------------------------------------------------------------
(...)
ValueError: Length mismatch: Expected axis has 2 elements, new values have 4 elements
```
```
In [61]: df = pd.DataFrame({"A": [0, 0, 7, np.nan], "B": [1, 2, 4, np.nan]}, index=list('abcd')) 
    ...: gb = df.groupby('B', dropna=False, axis=1) 
    ...: gb[['B']].transform(len)                                                                                                                                                                                 
---------------------------------------------------------------------------
(...)
IndexError: indices are out-of-bounds
```
```
In [62]: df = pd.DataFrame({"A": [0, 0, 7, np.nan], "B": [1, 2, 4, np.nan]}, index=list('abcd')) 
    ...: gb = df.groupby('B', dropna=True, axis=1) 
    ...: gb[['B']].transform(len)                                                                                                                                                                                 
---------------------------------------------------------------------------
(...)
ValueError: No objects to concatenate
```

Reply:
For the series case part of the problem is that `axis` does not propagate to `_set_result_index_ordered` (if I print `self.axis` inside the function it's always `0` even though it's set to 1 in the `groupby` call). For `DataFrameGroupBy` the errors happen prior to calling `_set_result_index_ordered`, though.

Would these be the todo's for moving this forward?

- [ ] add tests (xfail them for the moment)
- [ ] pass axis parameter to `_set_result_index_ordered` 
- [ ] investigate DataFrameGroupBy failures

Also, are you happy with addressing these here, or would you rather I open an issue & a follow-up PR?

Reply:
@rhshadrach noticed you already opened #36308 for the DataFrameGroupBy issue. I've opened  #36321 for the SeriesGroupBy part